repl/src/main/scala/com/cloudera/livy/repl/Session.scala

```diff
@@ -82,21 +84,53 @@ class Session(interpreter: Interpreter, stateChangedCallback: SessionState => Un
   def execute(code: String): Int = {
     val statementId = newStatementId.getAndIncrement()
     _statements(statementId) = new Statement(statementId, StatementState.Waiting, null)
+
     Future {
-      _statements(statementId) = new Statement(statementId, StatementState.Running, null)
+      setJobGroup(statementId)
+      _statements(statementId).state.compareAndSet(StatementState.Waiting, StatementState.Running)
+
+      val executeResult = if (_statements(statementId).state.get() != StatementState.Cancelling) {
+        executeCode(statementId, code)
+      } else {
+        null
+      }
 
-      _statements(statementId) =
-        new Statement(statementId, StatementState.Available, executeCode(statementId, code))
+      _statements(statementId).output = executeResult
+      _statements(statementId).state.compareAndSet(StatementState.Running, StatementState.Available)
+      _statements(statementId).state.compareAndSet(
+        StatementState.Cancelling, StatementState.Cancelled)
     }
+
     statementId
   }
 
+  def cancel(statementId: Int): Unit = {
+    if (_statements.contains(statementId)) {
+      val oldStmtState = _statements(statementId).state.getAndSet(StatementState.Cancelling)
+
+      info(s"Statement $statementId is cancelling")
+
+      val singleThreadExecutor = ExecutionContext.fromExecutorService(
+        Executors.newSingleThreadExecutor())
+      Future {
+        if (oldStmtState == StatementState.Running) {
+          while (_statements(statementId).state.get() != StatementState.Cancelled) {
```

Comment:
If the statement being cancelled doesn't make any calls to Spark, we can't cancel it.